expression/src/main/java/io/smallrye/common/expression/Expression.java

```diff
@@ -406,10 +409,19 @@ private static Node parseString(Itr itr, final boolean allowExpr, final boolean
                             //throw Assert.unreachableCode();
                         }
                         case '$': {
+                            if (idx > start) {
+                                // TP 4
+                                list.add(new LiteralNode(itr.getStr(), start, idx - 1));
+                            }
                             // $$
                             if (flags.contains(Flag.MINI_EXPRS)) {
                                 // TP 13
                                 list.add(new ExpressionNode(false, LiteralNode.DOLLAR, Node.NULL));
+                            } else if (flags.contains(Flag.NO_$$)) {
+                                list.add(LiteralNode.DOLLAR);
+                                if (!itr.hasNext() || itr.peekNext() != '{') {
+                                    itr.prev();
+                                }
```

Comment:
We shouldn't need to care whether or not the next character is a `{` for this case. We should be acting in exactly the same way as the default branch here if `MINI_EXPRS` is off and `NO_$$` is present.

Reply:
My idea was to still keep `$$` as an escape sequence in the presence of an expression, as:

`$${foo}` -> `${foo}`
`$$foo` -> `$$foo`

The goal is to minimize breaking user configuration

Reply:
I'm not sure that goal is as important as having a consistent parsing strategy. Whenever you make an exception case to escaping, then something breaks. You can either have it be something common but easy to fix, or rarer but harder to fix. If your escaping scheme is not general, then you have to hunt harder for the case that is not easy to fix but it's going to be there.

I think it's best to make the scheme as simple as possible, and just accept the breakage. We could potentially come up with a config migration scheme if needed, though I doubt it would be needed.

Reply:
Ok, so if we drop `$$` as an escape for this particular case, do we use `\$`?

Reply:
Yes, the new policy would be "`\` always escapes the character that comes after it; if you need to pass a `\` to the config layer then use `\\`". Then the only effect of `NO_$$` is that `$$` is not treated as a special sequence in any way unless `MINI_EXPRS` is turned on.

Reply:
Adding `ESCAPES`? Or we always add the `\$` escape to `NO_$$`?

By moving to `NO_$$`we are adding a braking change for configuration that relies on `$$` as the escape, which needs to change to `\$`. If we move to `ESCAPES`, this will also break comma escapes in comma-separated values.

I would prefer to keep the `ESCAPES` off and just add `\$` to `NO_$$`.

Reply:
If we add `\$` to `NO_$$` then we have a situation where `\` does different things based on context, and in order to get a `\` you'd need to do `\\` or something similar, and now you basically have an inconsistent version of `ESCAPES` because if you need `\` you say `\`, but if you need `\\` you have to say `\\\\`.

If, in config, we use `ESCAPES` and the simplified `NO_$$`, now we're 100% consistent: `\\` means `\`, `\$` means `$`, `\` followed by *anything* always means *something*, and `$` is an expression iff it is followed by `{`.